### PR TITLE
Dev UI - add basic monitoring of CDI components 

### PR DESCRIPTION
**Pull request by mkouba**
- by default, fired events and business method invocations are monitored
- `quarkus.arc.dev-mode-monitoring=false` disables the feature

### CODE REVIEW

File: extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/devconsole/EventsMonitor.java
Comment (gsmet) [inline extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/devconsole/EventsMonitor.java:20]: Is it really true? What if you disable the bean removal feature?

Wouldn't it make more sense to remove the annotation and declare the bean with a build item as we do everywhere else?

**Review by gsmet — DISMISSED** [review]
Added a small comment/question.

File: extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/devconsole/EventsMonitor.java
Comment (mkouba) [reply extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/devconsole/EventsMonitor.java:20]: It is true because of https://github.com/quarkusio/quarkus/pull/15040/files#diff-164fa9cb0c9912604128d33ae6c38003a72dde1fef0e77ac5d76af0ee7ee4087R46

The bean removal feature has no effect here.

Comment (gsmet) [reply extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/devconsole/EventsMonitor.java:20]: Ah sorry, I misread it as an `UnremovableBeanBuildItem`. High time I start my week-end!

File: extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/devconsole/InvocationInterceptor.java
Comment (manovotn) [inline extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/devconsole/InvocationInterceptor.java:43]: At this point, a user could have disabled/restarted request context. You should probably re-check if it's enabled.

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/devconsole/DevConsoleProcessor.java
Comment (manovotn) [inline extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/devconsole/DevConsoleProcessor.java:73]: Personally I'd keep them as it may be valuable in getting a complete picture of what's happening.

Alternatively, there could be yet another config option that would filter them out?

**Review by manovotn — COMMENTED** [review]
I played with this on some quickstarts and I think it is a great tool! Also, using CDI to monitor CDI is brilliant ;-)
Left just two minor comments, otherwise I am simply +1 for this.

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/devconsole/DevConsoleProcessor.java
Comment (mkouba) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/devconsole/DevConsoleProcessor.java:73]: I'll just remove this TODO... I don't think we need more config options now ;-)

File: extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/devconsole/InvocationInterceptor.java
Comment (mkouba) [reply extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/devconsole/InvocationInterceptor.java:43]: Hm, that's a good point. I think that there should be also a `@PreDestroy` callback on the invocation tree to flush the data if a request context is terminated unexpectedly...

Comment (mkouba) [reply extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/devconsole/InvocationInterceptor.java]: Ok, the current version should be more robust, although not perfect. But it's probably the best we can do if someone touches the request context manually...

**Review by gsmet — CHANGES_REQUESTED** [review]
I played a bit with it with my Quarkus Bot Java app.

A couple of things:
- fired events are numbered from 1. but they are sorted in anti-chronological order. So every time you fire a new event, the events are renumbered, which is weird.
- for invocation trees, I think you should had % based width for the table. Otherwise, every time I click on the + icon in the tree, the width changes, which is a bit annoying.
- still invocation trees, I find it weird you don't sum up the invocation times of the tree somehow. Maybe you should have two values. See screenshot.
Screenshot from 2021-02-18 13-54-39

**Comment by mkouba** [comment]
> fired events are numbered from 1. but they are sorted in anti-chronological order. So every time you fire a new event, the events are renumbered, which is weird.

Yes, the latest events/invocations are listed first and the number is just an index, a position in the table. We could just skip this column but I always find it useful when displaying a table...

> for invocation trees, I think you should had % based width for the table. Otherwise, every time I click on the + icon in the tree, the width changes, which is a bit annoying.

Ok, I can try do something with this but I'm no CSS master ;-). (it looks ok in FF BTW)

> still invocation trees, I find it weird you don't sum up the invocation times of the tree somehow. Maybe you should have two values.

Hm, that could be a bug with CP or something like that. The duration of the root must always be greater than the sum of the children.

**Comment by gsmet** [comment]
> Yes, the latest events/invocations are listed first and the number is just an index, a position in the table. We could just skip this column but I always find it useful when displaying a table...

It's useful if it's stable. If it changes every time we get an event, I think it's not such a good idea. Can't you have 1. for the oldest event instead? This way, you would have a number and something that's stable.

If not, I would drop it.

> Ok, I can try do something with this but I'm no CSS master ;-). (it looks ok in FF BTW)

I think just adding `style="width:xx%"` to the ``th``s should be enough. If not I can have a look.

> Hm, that could be a bug with CP or something like that. The duration of the root must always be greater than the sum of the children.

Well, you have seen my screenshot :). Who could help on this?

**Comment by mkouba** [comment]
@gsmet all your commets should be addressed now ;-)

**Review by manovotn — APPROVED** [review]
I am just putting my green stamp here as all my previous comments were addressed.
But we should still wait until Guillaume gives his approval.

**Review by gsmet — APPROVED** [review]
Woot, thanks! That's very useful!

**Pull request merged**
